twine/package.py

```diff
@@ -133,104 +133,51 @@ def from_filename(cls, filename: str, comment: Optional[str]) -> "PackageFile":
                 "Unknown distribution format: '%s'" % os.path.basename(filename)
             )
 
-        supported_metadata = list(pkginfo.distribution.HEADER_ATTRS)
-        if cls._is_unknown_metadata_version(captured):
+        # Parse and validate metadata.
+        meta, unparsed = metadata.parse_email(data)
+        if unparsed:
             raise exceptions.InvalidDistribution(
-                "Make sure the distribution is using a supported Metadata-Version: "
-                f"{', '.join(supported_metadata)}."
+                "Invalid distribution metadata: {}".format(
+                    ", ".join("unrecognized or malformed field {key!r}")
+                    for key in unparsed
+                )
             )
-        # If pkginfo <1.11 encounters a metadata version it doesn't support, it may give
-        # back empty metadata. At the very least, we should have a name and version,
-        # which could also be empty if, for example, a MANIFEST.in doesn't include
-        # setup.cfg.
-        missing_fields = [
-            f.capitalize() for f in ["name", "version"] if not getattr(meta, f)
```

Comment:
Here's where I see we lost capitalization (and canonical names) of the fields missing in the message.

Reply:
Not exactly. The change in error messages are a consequence of using `packaging.metadata` for metadata parsing and validation.